### PR TITLE
Don't switch streams when running Caffe2 ops from c10.

### LINKED ISSUE #54830: CUDAContext::SwitchToDevice() acquires new CUDA stream instead of using the default stream
## 🐛 Bug

The first call to `caffe2::CUDAContext::SwitchToDevice(0)` will ignore the default CUDA Stream and acquire a new low priority CUDA Stream and set this stream as the new current stream for the given device. 

## To Reproduce

1. Run trace that allocates memory using the default stream and subsequently calls into `caffe2::CUDAContext::SwitchToDevice()`
2. Run the trace again any amount of times
3. Observe that the subsequent runs of the trace allocate memory from a stream that is not the default stream. CUDACachingAllocator is not able to re-use the memory allocated from the first execution because it was allocated from a separate stream.

Here is the portion of the stack trace in question:
```
caffe2::ThreadLocalCUDAObjects::GetCUDAStream(c10::StreamId stream_id, c10::DeviceIndex gpu, caffe2::ThreadLocalCUDAObjects * const this)
caffe2::ThreadLocalCUDAObjects::SetCurrentStreamId(c10::StreamId stream_id, c10::DeviceIndex gpu, caffe2::ThreadLocalCUDAObjects * const this) 
caffe2::CUDAContext::SwitchToDevice(c10::StreamId stream_id, caffe2::CUDAContext * const this) 
at::BaseContext::SwitchToDevice(caffe2::CUDAContext * const this)
```

## Expected behavior

Calls to `caffe2::CUDAContext::SwitchToDevice()` should use the default stream and not create a new stream. 

## Environment

```
PyTorch version: 1.7.0
Is debug build: False
CUDA used to build PyTorch: 10
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.14.0

Python version: 3.6 (64-bit runtime)
Is CUDA available: N/A
CUDA runtime version: 11.1.74
GPU models and configuration: GPU 0: GeForce RTX 2080 Ti
Nvidia driver version: 455.23.04
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[conda] Could not collect

```

## Additional context

The constructor of `caffe2::ThreadLocalCUDAObjects` initializes a vector of CUDA stream for each GPU in the `cuda_streams_` member variable. These vectors are all initialized to empty and populated on subsequent calls to `caffe2::ThreadLocalCUDAObjects::GetCUDAStream`. One such code path is from the call stack shown above. Specifically, `caffe2::CUDAContext::SwitchToDevice(0)` is called from [context_base.h#L48](https://github.com/pytorch/pytorch/blob/8c798e062216278673a75bac0848ea69a8bd3f03/caffe2/core/context_base.h#L48) (via [context_gpu.h#L180](https://github.com/pytorch/pytorch/blob/8c798e062216278673a75bac0848ea69a8bd3f03/caffe2/core/context_gpu.h#L180)) with the apparent intention to use the default stream (based on notes in PR #15126). However, this code path triggers a call to `c10::cuda::getStreamFromPool` from within `caffe2::ThreadLocalCUDAObjects::GetCUDAStream` due to the fact that the `caffe2::ThreadLocalCUDAObjects::cuda_streams_` member has no reference to the default stream. This causes a new stream to be acquired (as per [tensor_cuda_stream](https://github.com/pytorch/pytorch/blob/8c798e062216278673a75bac0848ea69a8bd3f03/docs/cpp/source/notes/tensor_cuda_stream.rst) documentation).

It seems that either the `cuda_streams_` member variable should be initialized to include the default stream, or `GetCUDAStream` should check for logical stream id 0 and call `getDefaultCUDAStream` instead of `getStreamFromPool`.

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55121 Don't switch streams when running Caffe2 ops from c10.**

This is done by allow -1 as a stream ID, meaning "don't change
the stream", in SwitchToDevice

Fixes #54830

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27527544](https://our.internmc.facebook.com/intern/diff/D27527544)

### CODE REVIEW

**Comment by ezyang** [comment]
cc @jwestell, see if this fixes your problem

File: caffe2/core/export_caffe2_op_to_c10.h [hunk 1]
```diff
@@ -29,7 +29,7 @@ inline c10::List<at::Tensor> _call_caffe2_op(
     std::vector<c10::IValue>&& inputs,
     c10::List<at::Tensor>&& outputs) {
   Caffe2Operator op(schema, std::move(inputs), std::move(outputs));
-  op.Run();
+  op.Run(-1);
```
Comment (jwestell) [inline caffe2/core/export_caffe2_op_to_c10.h:32]: @ezyang, I believe it is the line above this (constructor of `caffe2::Operator` [[operator.h#L814](https://github.com/pytorch/pytorch/blob/7824d8277aebc20ac8c21c5f4b52a29182c24816/caffe2/core/operator.h#L814)] that does the initial switching to device, and thus the additional stream is already created and switched to here.

Comment (ezyang) [reply caffe2/core/export_caffe2_op_to_c10.h:32]: Good catch.

**Comment by jwestell** [comment]
The patch works great, @ezyang. My application stays on the defualt stream after using the caffe2 op. Thanks for your help.
FWIW, I tested this patch backported to the v1.7.0 tag.

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**